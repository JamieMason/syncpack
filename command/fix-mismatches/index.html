<!doctype html><title>Redirecting to: /command/fix</title><meta http-equiv="refresh" content="0;url=/command/fix"><meta name="robots" content="noindex"><link rel="canonical" href="https://jamiemason.github.io/command/fix"><body>	<a href="/command/fix">Redirecting from <code>/syncpack/command/fix-mismatches/</code> to <code>/command/fix</code></a></body>