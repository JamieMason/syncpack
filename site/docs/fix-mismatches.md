---
id: fix-mismatches
title: fix-mismatches
---
