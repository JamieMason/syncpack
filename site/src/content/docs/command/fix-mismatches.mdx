---
title: fix-mismatches
---

import ConfigOption from '@site/partials/option/_config.mdx';
import FilterOption from '@site/partials/option/_filter.mdx';
import HelpOption from '@site/partials/option/_help.mdx';
import IndentOption from '@site/partials/option/_indent.mdx';
import SourceOption from '@site/partials/option/_source.mdx';
import SpecsOption from '@site/partials/option/_specs.mdx';
import TypesOption from '@site/partials/option/_types.mdx';
import LogLevelOption from '@site/partials/option/_log-level.mdx';

Ensure that multiple packages requiring the same dependency define the same version, so that every package requires eg. `react@16.4.2`, instead of a combination of `react@16.4.2`, `react@0.15.9`, and `react@16.0.0`.

## CLI Options

### source

<SourceOption command="fix-mismatches" />

### filter

<FilterOption command="fix-mismatches" />

### config

<ConfigOption command="fix-mismatches" />

### specs

<SpecsOption command="fix-mismatches" />

### types

<TypesOption command="fix-mismatches" />

### indent

<IndentOption command="fix-mismatches" />

### help

<HelpOption command="fix-mismatches" />

## Troubleshooting

<LogLevelOption command="fix-mismatches" />
