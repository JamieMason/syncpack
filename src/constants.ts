import type { RcConfig, SemverRange } from './config/types.js';

/** Single source of truth, intended to aid testing or to override */
export const CWD = process.env.MOCK_CWD || process.cwd();

/** Single source of truth for icons used in output */
export const ICON = {
  banned: '⦸',
  cross: '✘',
  debug: '?',
  info: 'i',
  panic: '!',
  rightArrow: '→',
  skip: '-',
  tick: '✓',
} as const;

export const RANGE = {
  ANY: '*',
  EXACT: '',
  GT: '>',
  GTE: '>=',
  LOOSE: '.x',
  LT: '<',
  LTE: '<=',
  MINOR: '^',
  PATCH: '~',
  WORKSPACE: 'workspace:',
} as const satisfies Record<string, SemverRange>;

export const INTERNAL_TYPES = [
  'dev',
  'local',
  'overrides',
  'peer',
  'pnpmOverrides',
  'prod',
  'resolutions',
] as const;

export const CUSTOM_TYPES = {
  dev: {
    strategy: 'versionsByName',
    path: 'devDependencies',
  },
  local: {
    strategy: 'name~version',
    namePath: 'name',
    path: 'version',
  },
  overrides: {
    strategy: 'versionsByName',
    path: 'overrides',
  },
  peer: {
    strategy: 'versionsByName',
    path: 'peerDependencies',
  },
  pnpmOverrides: {
    strategy: 'versionsByName',
    path: 'pnpm.overrides',
  },
  prod: {
    strategy: 'versionsByName',
    path: 'dependencies',
  },
  resolutions: {
    strategy: 'versionsByName',
    path: 'resolutions',
  },
} as const;

export const DEFAULT_CONFIG = {
  customTypes: CUSTOM_TYPES,
  dependencyTypes: ['**'],
  filter: '.',
  formatBugs: true,
  formatRepository: true,
  indent: '  ',
  lintFormatting: true,
  lintSemverRanges: true,
  lintVersions: true,
  semverGroups: [],
  sortAz: [
    'bin',
    'contributors',
    'dependencies',
    'devDependencies',
    'keywords',
    'peerDependencies',
    'resolutions',
    'scripts',
  ],
  sortExports: [
    'types',
    'node-addons',
    'node',
    'browser',
    'import',
    'require',
    'development',
    'production',
    'default',
  ],
  sortFirst: ['name', 'description', 'version', 'author'],
  sortPackages: true,
  source: ['package.json', 'packages/*/package.json'],
  specifierTypes: ['**'],
  versionGroups: [],
} satisfies RcConfig;
