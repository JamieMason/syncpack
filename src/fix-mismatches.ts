import chalk from 'chalk';
import { writeJson } from 'fs-extra';
import * as _ from 'lodash';
import { relative } from 'path';
import {
  OPTION_INDENT,
  OPTION_SOURCES,
  OPTIONS_DEV,
  OPTIONS_FILTER_DEPENDENCIES,
  OPTIONS_PEER,
  OPTIONS_PROD
} from './constants';
import { collect } from './lib/collect';
import { getDependencyTypes } from './lib/get-dependency-types';
import { getIndent } from './lib/get-indent';
import { getPackages } from './lib/get-packages';
import { getMismatchedVersionsByName } from './lib/get-versions-by-name';
import { getNewest } from './lib/version';
import { CommanderApi, IManifest } from './typings';

export const run = async (program: CommanderApi) => {
  program
    .option(OPTION_SOURCES.spec, OPTION_SOURCES.description, collect)
    .option(OPTIONS_PROD.spec, OPTIONS_PROD.description)
    .option(OPTIONS_DEV.spec, OPTIONS_DEV.description)
    .option(OPTIONS_PEER.spec, OPTIONS_PEER.description)
    .option(OPTION_INDENT.spec, OPTION_INDENT.description)
    .option(
      OPTIONS_FILTER_DEPENDENCIES.spec,
      OPTIONS_FILTER_DEPENDENCIES.description
    )
    .parse(process.argv);

  const pkgs = getPackages(program);
  const dependencyTypes = getDependencyTypes(program);
  const indent = getIndent(program);
  const mismatchedVersionsByName = getMismatchedVersionsByName(
    dependencyTypes,
    pkgs,
    program.filter
  );

  await Promise.all(
    pkgs.map(({ data, path }) => {
      const nextData: IManifest = JSON.parse(JSON.stringify(data));
      const shortPath = `./${relative('.', path)}`;
      const changes: string[][] = [];
      Object.entries(mismatchedVersionsByName).forEach(([name, versions]) => {
        const newest = getNewest(versions);
        dependencyTypes.forEach((type) => {
          if (
            nextData[type] &&
            nextData[type][name] &&
            nextData[type][name] !== newest
          ) {
            changes.push([name, nextData[type][name], newest]);
            nextData[type][name] = newest;
          }
        });
      });
      if (changes.length > 0) {
        changes.forEach(([name, from, to]) => {
          console.log(
            chalk.bgYellow.black(' FIXED '),
            chalk.blue(shortPath),
            name,
            chalk.red(from),
            'â†’',
            chalk.green(to)
          );
        });
        return writeJson(path, nextData, { spaces: indent });
      }
      console.log(chalk.bgGreen.black(' VALID '), chalk.blue(shortPath));
    })
  );
};
