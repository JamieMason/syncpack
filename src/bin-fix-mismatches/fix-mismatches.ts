import chalk from 'chalk-template';
import { Context, Effect, flow, pipe } from 'effect';
import { isObject } from 'tightrope/guard/is-object.js';
import { isUndefined } from 'tightrope/guard/is-undefined.js';
import { logIgnoredSize } from '../bin-lint-semver-ranges/lint-semver-ranges.js';
import {
  logMissingLocalVersion,
  logMissingSnappedToMismatch,
  logSameRangeMismatch,
  logUnsupportedMismatch,
} from '../bin-list-mismatches/list-mismatches.js';
import { CliConfigTag } from '../config/tag.js';
import { type CliConfig } from '../config/types.js';
import { ICON } from '../constants.js';
import type { ErrorHandlers } from '../error-handlers/default-error-handlers.js';
import { defaultErrorHandlers } from '../error-handlers/default-error-handlers.js';
import type { Ctx } from '../get-context/index.js';
import { getContext } from '../get-context/index.js';
import { getInstances } from '../get-instances/index.js';
import { exitIfInvalid } from '../io/exit-if-invalid.js';
import type { Io } from '../io/index.js';
import { IoTag } from '../io/index.js';
import { writeIfChanged } from '../io/write-if-changed.js';
import { getVersionGroupHeader } from '../lib/get-group-header.js';
import { padStart } from '../lib/pad-start.js';
import { withLogger } from '../lib/with-logger.js';
import type { Report } from '../report.js';
import { DELETE } from '../version-group/lib/delete.js';

interface Input {
  io: Io;
  cli: Partial<CliConfig>;
  errorHandlers?: ErrorHandlers;
}

export function fixMismatches({ io, cli, errorHandlers = defaultErrorHandlers }: Input) {
  return pipe(
    getContext({ io, cli, errorHandlers }),
    Effect.flatMap((ctx) =>
      pipe(
        Effect.gen(function* ($) {
          const { versionGroups } = yield* $(getInstances(ctx, io, errorHandlers));
          let index = 0;
          for (const group of versionGroups) {
            const groupSize = group.instances.length;
            let fixedCount = 0;
            let unfixableCount = 0;
            let validCount = 0;

            if (group._tag === 'FilteredOut') {
              index++;
              continue;
            }

            yield* $(Effect.logInfo(getVersionGroupHeader({ group, index })));

            if (group._tag === 'Ignored') {
              yield* $(logIgnoredSize(groupSize));
              index++;
              continue;
            }

            for (const groupReport of yield* $(group.inspectAll())) {
              for (const report of groupReport.reports) {
                if (report._tagGroup === 'Valid') {
                  validCount++;
                } else if (report._tagGroup === 'Fixable') {
                  fixedCount++;
                  yield* $(fixMismatch(report));
                } else if (report._tagGroup === 'Unfixable') {
                  ctx.isInvalid = true;
                  unfixableCount++;
                }

                if (report._tag === 'MissingLocalVersion') {
                  yield* $(logMissingLocalVersion(report));
                } else if (report._tag === 'MissingSnappedToMismatch') {
                  yield* $(logMissingSnappedToMismatch(report));
                } else if (report._tag === 'UnsupportedMismatch') {
                  yield* $(logUnsupportedMismatch(report));
                } else if (report._tag === 'SameRangeMismatch') {
                  yield* $(logSameRangeMismatch(report));
                }
              }
            }

            if (validCount) yield* $(logAlreadyValidSize(validCount));
            if (fixedCount) yield* $(logFixedSize(fixedCount));
            if (unfixableCount) yield* $(logUnfixableSize(unfixableCount));

            index++;
          }

          yield* $(removeEmptyObjects(ctx));

          return ctx;
        }),
        Effect.flatMap(writeIfChanged),
        Effect.catchTags({
          WriteFileError: flow(
            errorHandlers.WriteFileError,
            Effect.map(() => {
              ctx.isInvalid = true;
              return ctx;
            }),
          ),
        }),
        Effect.flatMap(exitIfInvalid),
      ),
    ),
    Effect.provide(pipe(Context.empty(), Context.add(CliConfigTag, cli), Context.add(IoTag, io))),
    withLogger,
  );
}

export function fixMismatch(report: Report.Version.Fixable.Any) {
  return report.fixable.instance.write(report._tag === 'Banned' ? DELETE : report.fixable.raw);
}

/** Remove empty objects such as `{"dependencies": {}}` left after deleting */
function removeEmptyObjects(ctx: Ctx) {
  return Effect.sync(() => {
    ctx.packageJsonFiles.forEach((file) => {
      const contents = file.jsonFile.contents;
      Object.keys(contents).forEach((key) => {
        const value = contents[key];
        if (isObject(value) && value && Object.values(value).every(isUndefined)) {
          delete contents[key];
        }
      });
    });
  });
}

export function logAlreadyValidSize(amount: number) {
  const msg = chalk`${padStart(amount)} {green ${ICON.tick}} already valid`;
  return Effect.logInfo(msg);
}

export function logFixedSize(amount: number) {
  const msg = chalk`${padStart(amount)} {green ${ICON.tick}} fixed`;
  return Effect.logInfo(msg);
}

export function logUnfixableSize(amount: number) {
  const msg = chalk`{red ${padStart(amount)} ${
    ICON.panic
  } can be fixed manually using} {blue syncpack prompt}`;
  return Effect.logInfo(msg);
}
