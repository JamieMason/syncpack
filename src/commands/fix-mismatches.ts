import chalk from 'chalk';
import { writeFileSync } from 'fs-extra';
import { EOL } from 'os';
import { relative } from 'path';
import { CWD, SyncpackConfig } from '../constants';
import { getHighestVersion } from './lib/get-highest-version';
import { getWrappers, SourceWrapper } from './lib/get-wrappers';
import { getMismatchedDependencies } from './lib/installations/get-mismatched-dependencies';
import { log } from './lib/log';
import { matchesFilter } from './lib/matches-filter';

type Options = Pick<SyncpackConfig, 'dev' | 'filter' | 'indent' | 'peer' | 'prod' | 'source' | 'versionGroups' | 'versionOverrides'>;

const getWorkspaceVersion = (name: string, wrappers: SourceWrapper[]): string | null => {
  const local = wrappers.find((wrapper) => wrapper.contents.name === name);
  return local?.contents?.version || null;
};

const getOverridesVersion = (versionOverrides: SyncpackConfig['versionOverrides'], packageName: string) => versionOverrides?.[packageName] || null;

export const fixMismatches = (wrappers: SourceWrapper[], options: Options): void => {
  const iterator = getMismatchedDependencies(wrappers, options);
  const mismatches = Array.from(iterator).filter(matchesFilter(options));

  mismatches.forEach((installedPackage) => {
    const versions = installedPackage.installations.map((installation) => installation.version);
    const nextVersion =
      getOverridesVersion(options.versionOverrides, installedPackage.name) ||
      getWorkspaceVersion(installedPackage.name, wrappers) ||
      getHighestVersion(versions);
    if (nextVersion !== null) {
      installedPackage.installations.forEach(({ type, name, source }) => {
        const dependencies = source.contents[type];
        if (dependencies) {
          dependencies[name] = nextVersion;
        }
      });
    }
  });
};

export const fixMismatchesToDisk = (options: Options): void => {
  const { indent, source } = options;
  const toJson = (wrapper: SourceWrapper): string => `${JSON.stringify(wrapper.contents, null, indent)}${EOL}`;
  const wrappers = getWrappers({ source });
  const allBefore = wrappers.map(toJson);

  fixMismatches(wrappers, options);

  wrappers.forEach((wrapper, i) => {
    const shortPath = relative(CWD, wrapper.filePath);
    const before = allBefore[i];
    const after = toJson(wrapper);
    if (before !== after) {
      writeFileSync(wrapper.filePath, after);
      log(chalk.green('âœ“'), shortPath);
    } else {
      log(chalk.dim('-', shortPath));
    }
  });
};
