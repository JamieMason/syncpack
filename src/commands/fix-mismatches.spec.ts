import * as mock from '../../test/mock';
import { DEFAULT_CONFIG } from '../constants';
import * as api from './fix-mismatches';

describe('fixMismatches', () => {
  let fixMismatches: typeof api.fixMismatches;

  afterEach(() => {
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.mock('./lib/log', () => ({ log: jest.fn() }));
    fixMismatches = require('./fix-mismatches').fixMismatches;
  });

  describe('when dependencies are installed with different versions', () => {
    describe('when the dependency is not a package maintained in this workspace', () => {
      describe('when strict overrides are provided', () => {
        it('uses the version from the versionOverrides map', () => {
          const wrappers = [mock.wrapper('a', ['foo@0.1.0']), mock.wrapper('b', [], ['foo@0.2.0'])];
          fixMismatches(wrappers, {...DEFAULT_CONFIG, versionOverrides: { 'foo': '0.3.0' }});
          expect(wrappers).toMatchSnapshot();
        })
      });

      it('uses the highest version', () => {
        const wrappers = [mock.wrapper('a', ['foo@0.1.0']), mock.wrapper('b', [], ['foo@0.2.0'])];
        fixMismatches(wrappers, DEFAULT_CONFIG);
        expect(wrappers).toMatchSnapshot();
      });
    });

    describe('when the dependency is a package maintained in this workspace', () => {
      it('uses the workspace version', () => {
        const wrappers = [
          mock.wrapper('a', ['foo@0.1.0']),
          mock.wrapper('b', [], ['foo@0.2.0']),
          mock.wrapper('foo', [], [], [], { name: 'foo', version: '0.0.1' }),
        ];
        fixMismatches(wrappers, DEFAULT_CONFIG);
        expect(wrappers).toMatchSnapshot();
      });
    });

    it('replaces non-semver dependencies with valid semver dependencies', () => {
      const wrappers = [
        mock.wrapper('a', ['foo@link:vendor/foo-0.1.0']),
        mock.wrapper('b', ['foo@link:vendor/foo-0.2.0']),
        mock.wrapper('c', ['foo@0.3.0']),
        mock.wrapper('d', ['foo@0.2.0']),
      ];
      fixMismatches(wrappers, DEFAULT_CONFIG);
      expect(wrappers).toMatchSnapshot();
    });
  });
});
