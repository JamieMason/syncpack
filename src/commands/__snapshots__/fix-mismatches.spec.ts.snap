// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixMismatches when dependencies are installed with different versions replaces non-semver dependencies with valid semver dependencies 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/a/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"link:vendor/foo-0.1.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/b/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"link:vendor/foo-0.2.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/c/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"0.3.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/d/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"0.2.0\\"
  }
}
",
  },
]
`;

exports[`fixMismatches when dependencies are installed with different versions when the dependency is a package maintained in this workspace uses the workspace version 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.0.1",
      },
    },
    "filePath": "/a/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"0.1.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "devDependencies": Object {
        "foo": "0.0.1",
      },
    },
    "filePath": "/b/package.json",
    "json": "{
  \\"devDependencies\\": {
    \\"foo\\": \\"0.2.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "name": "foo",
      "version": "0.0.1",
    },
    "filePath": "/foo/package.json",
    "json": "{
  \\"name\\": \\"foo\\",
  \\"version\\": \\"0.0.1\\"
}
",
  },
]
`;

exports[`fixMismatches when dependencies are installed with different versions when the dependency is not a package maintained in this workspace uses the highest version 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.2.0",
      },
    },
    "filePath": "/a/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"0.1.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "devDependencies": Object {
        "foo": "0.2.0",
      },
    },
    "filePath": "/b/package.json",
    "json": "{
  \\"devDependencies\\": {
    \\"foo\\": \\"0.2.0\\"
  }
}
",
  },
]
`;

exports[`fixMismatches when dependencies are installed with different versions when the dependency is not a package maintained in this workspace when strict overrides are provided uses the version from the versionOverrides map 1`] = `
Array [
  Object {
    "contents": Object {
      "dependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/a/package.json",
    "json": "{
  \\"dependencies\\": {
    \\"foo\\": \\"0.1.0\\"
  }
}
",
  },
  Object {
    "contents": Object {
      "devDependencies": Object {
        "foo": "0.3.0",
      },
    },
    "filePath": "/b/package.json",
    "json": "{
  \\"devDependencies\\": {
    \\"foo\\": \\"0.2.0\\"
  }
}
",
  },
]
`;
